{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "758edd83-be47-4558-b5bb-d3cc58ff2965",
   "metadata": {},
   "source": [
    "# General genomic information\n",
    "Author: Thomas Cokelaer, Institut Pasteur, ParSig team, 2025"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6b56bdd4-46f5-48f3-9758-8600d4dcd6c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sequana import GFF3, FastA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0570554-bd6d-4a32-8f59-c2f93dabecf9",
   "metadata": {},
   "source": [
    "## Infantum genome size and  number of coding genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6a8cc31e-575c-4ebc-8996-b8d81c860eb2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes in infantum=8683\n",
      "Total length=32.803248Mb\n"
     ]
    }
   ],
   "source": [
    "g = GFF3(\"references/infantum.gff\")\n",
    "N = len(g.df.query(\"genetic_type=='gene'\"))\n",
    "print(f\"Number of genes in infantum={N}\")\n",
    "L = g.df.query(\"genetic_type=='region'\")[['start', 'stop', 'Name']].dropna().stop.sum()/1e6\n",
    "print(f\"Total length={L}Mb\")\n",
    "assert N == 8683"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0239e315-b4fb-4338-a732-913948027c54",
   "metadata": {},
   "source": [
    "## Tarentolae genome size and  number of coding genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1dbdfcdf-8504-4095-b994-9d560514b347",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total length=35.416496Mb\n",
      "Number of genes in tarentolae=8703\n",
      "Total length ignoring unassigned contigs=32.283028Mb, ncontigs=57\n",
      "Number of genes in tarentolae ignoring unassigned contigs=8703\n"
     ]
    }
   ],
   "source": [
    "g = GFF3(\"references/tarentolae.gff\")\n",
    "L = g.df.query(\"genetic_type=='region'\")[['start', 'stop', 'Name']].stop.sum()/1e6\n",
    "print(f\"Total length={L}Mb\")\n",
    "\n",
    "N = len(g.df.query(\"genetic_type=='gene'\"))\n",
    "print(f\"Number of genes in tarentolae={N}\")\n",
    "\n",
    "L = g.df.query(\"genetic_type=='region'\")[['start', 'stop', 'Name']].dropna().stop.sum()/1e6\n",
    "nL = len(g.df.query(\"genetic_type=='region'\")[['start', 'stop', 'Name']].dropna().stop)\n",
    "print(f\"Total length ignoring unassigned contigs={L}Mb, ncontigs={nL}\")\n",
    "\n",
    "valid_contigs = g.df.loc[g.df.query(\"genetic_type=='region'\")[['start', 'stop', 'Name']].dropna().index].seqid.values\n",
    "\n",
    "N2 = len(g.df.query(\"genetic_type=='gene' and seqid in @valid_contigs\"))\n",
    "print(f\"Number of genes in tarentolae ignoring unassigned contigs={N}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e3a54f72-e01a-40c2-884d-9df38c6cf6ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "91.16068907088393"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ratio of genome within assigned contigs \n",
    "32.28/35.41 *100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1114b5a6-b41d-4449-975d-1ec8c6494d15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "97.2538205216592"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ratio of coding genes within assigned contigs (larger)\n",
    "N2/N *100"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03cc3c27-7e73-4eb8-98b5-3be2a259ea6f",
   "metadata": {},
   "source": [
    "## GC content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "922ea2ee-bd15-48c5-a744-4400867d4f59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "59.744678331852995"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = FastA(\"references/infantum.fa\")\n",
    "100 * sum([seq.upper().count('G') + seq.upper().count('C') for seq in f.sequences]) / f.get_stats()['total_length']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2da129be-c069-4d38-9f27-ae2610472635",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "57.41399431496555"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = FastA(\"references/tarentolae.fa\")\n",
    "100 * sum([seq.upper().count('G') + seq.upper().count('C') for seq in f.sequences]) / f.get_stats()['total_length']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3772e076-5108-40fe-bbac-fcd9628be653",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
