{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "42417e65-a499-4ae9-a934-9f6b4ff46437",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10954060-a5f8-4930-8342-208fa1637298",
   "metadata": {},
   "source": [
    "# Orthogroup construction between L. Tarentolae and L. infantum\n",
    "Author: Dr Thomas Cokelaer, 2025 \n",
    "\n",
    "- reference infantum: GCA_900500625.2   -- 36 contigs \n",
    "- reference tarentolae: GCA_009731335.1    -- 179 contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b95e6ac1-6f1b-4e3b-bf30-11cbbb2ad487",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sequana import *\n",
    "from pylab import *\n",
    "from bioconvert import fasta2faa # depends only on python\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7040405-4a72-4765-8947-9114a7f46854",
   "metadata": {},
   "source": [
    "# first, we need to build the orthogroups using the GFF and orthofinder. \n",
    "\n",
    "We need to extract the amino acids sequences (instead of DNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e655bfb9-a2f2-4676-9b89-92a60dae7c5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(\"analysis/orthodata/\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "adad58d4-2dd4-4a11-a4bb-311686186507",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bioconvert version 1.1 with _ for stop codon so that sequences are not stop in the middle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "180221ab-7275-4400-a329-cb28e4484b33",
   "metadata": {},
   "outputs": [],
   "source": [
    "if os.path.exists(\"references/proteins/tarentolae.faa\") is False:\n",
    "    g = GFF3(\"references/tarentolae.gff\")\n",
    "    g.save_gff_filtered(\"temp.gff\", features = ['gene'])\n",
    "    g.to_fasta(\"references/tarentolae.fa\", \"temp.fa\", features=['gene'], identifier=\"ID\")\n",
    "    conv = fasta2faa.FASTA2FAA(\"temp.fa\", \"references/proteins/tarentolae.faa\")    \n",
    "    conv()\n",
    "\n",
    "if os.path.exists(\"references/proteins/infantum.faa\") is False:\n",
    "    g = GFF3(\"references/infantum.gff\")\n",
    "    g.save_gff_filtered(\"temp.gff\", features = ['gene'])\n",
    "    g.to_fasta(\"references/infantum.fa\", \"temp.fa\", features=['gene'], identifier=\"ID\")\n",
    "    from bioconvert import fasta2faa # depends only on python\n",
    "    conv = fasta2faa.FASTA2FAA(\"temp.fa\", \"references/proteins/infantum.faa\")\n",
    "    conv()\n",
    "\n",
    "if os.path.exists(\"temp.fa\"):\n",
    "    os.unlink(\"temp.fa\")\n",
    "if os.path.exists(\"temp.gff\"):\n",
    "    os.unlink(\"temp.gff\")\n",
    "\n",
    "try:\n",
    "    os.symlink(\"references/proteins/infantum.faa\", \"analysis/orthodata/infantum.faa\")\n",
    "except:\n",
    "    pass\n",
    "\n",
    "\n",
    "try:\n",
    "    os.symlink(\"references/proteins/tarentolae.faa\", \"analysis/orthodata/tarentolae.faa\")\n",
    "except:\n",
    "    pass\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6f009326-feca-4b7b-aea0-7310aee33ad9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OrthoFinder version 2.5.5 Copyright (C) 2014 David Emms\n",
      "\n",
      "2026-01-16 22:08:45 : Starting OrthoFinder 2.5.5\n",
      "20 thread(s) for highly parallel tasks (BLAST searches etc.)\n",
      "2 thread(s) for OrthoFinder algorithm\n",
      "\n",
      "Checking required programs are installed\n",
      "----------------------------------------\n",
      "Test can run \"mcl -h\" - ok\n",
      "Test can run \"fastme -i TEMP/Results_Jan16/WorkingDirectory/dependencies/SimpleTest.phy -o TEMP/Results_Jan16/WorkingDirectory/dependencies/SimpleTest.tre\" - ok\n",
      "\n",
      "Dividing up work for BLAST for parallel processing\n",
      "--------------------------------------------------\n",
      "2026-01-16 22:08:45 : Creating diamond database 1 of 2\n",
      "2026-01-16 22:08:45 : Creating diamond database 2 of 2\n",
      "\n",
      "Running diamond all-versus-all\n",
      "------------------------------\n",
      "Using 20 thread(s)\n",
      "2026-01-16 22:08:45 : This may take some time....\n",
      "2026-01-16 22:09:31 : Done all-versus-all sequence search\n",
      "\n",
      "Running OrthoFinder algorithm\n",
      "-----------------------------\n",
      "2026-01-16 22:09:31 : Initial processing of each species\n",
      "2026-01-16 22:09:32 : Initial processing of species 0 complete\n",
      "2026-01-16 22:09:32 : Initial processing of species 1 complete\n",
      "2026-01-16 22:09:34 : Connected putative homologues\n",
      "2026-01-16 22:09:34 : Written final scores for species 0 to graph file\n",
      "2026-01-16 22:09:34 : Written final scores for species 1 to graph file\n",
      "2026-01-16 22:09:35 : Ran MCL\n",
      "\n",
      "Writing orthogroups to file\n",
      "---------------------------\n",
      "OrthoFinder assigned 16413 genes (94.4% of total) to 7627 orthogroups. Fifty percent of all genes were in orthogroups with 2 or more genes (G50 was 2) and were contained in the largest 3767 orthogroups (O50 was 3767). There were 7457 orthogroups with all species present and 7039 of these consisted entirely of single-copy genes.\n",
      "\n",
      "2026-01-16 22:09:36 : Done orthogroups\n",
      "\n",
      "Analysing Orthogroups\n",
      "=====================\n",
      "\n",
      "Calculating gene distances\n",
      "--------------------------\n",
      "2026-01-16 22:09:38 : Done\n",
      "2026-01-16 22:09:39 : Done 0 of 231\n",
      "2026-01-16 22:09:39 : Done 100 of 231\n",
      "2026-01-16 22:09:39 : Done 200 of 231\n",
      "\n",
      "Inferring gene and species trees\n",
      "--------------------------------\n",
      "\n",
      "Reconciling gene trees and species tree\n",
      "---------------------------------------\n",
      "2026-01-16 22:09:40 : Starting Recon and orthologues\n",
      "2026-01-16 22:09:40 : Starting OF Orthologues\n",
      "2026-01-16 22:09:40 : Done 0 of 231\n",
      "2026-01-16 22:09:40 : Done 100 of 231\n",
      "2026-01-16 22:09:40 : Done 200 of 231\n",
      "2026-01-16 22:09:41 : Done OF Orthologues\n",
      "\n",
      "Writing results files\n",
      "=====================\n",
      "2026-01-16 22:09:41 : Done orthologues\n",
      "\n",
      "Results:\n",
      "    TEMP/Results_Jan16/\n",
      "\n",
      "CITATION:\n",
      " When publishing work that uses OrthoFinder please cite:\n",
      " Emms D.M. & Kelly S. (2019), Genome Biology 20:238\n",
      "\n",
      " If you use the species tree in your work then please also cite:\n",
      " Emms D.M. & Kelly S. (2017), MBE 34(12): 3267-3278\n",
      " Emms D.M. & Kelly S. (2018), bioRxiv https://doi.org/10.1101/267914\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# run only once\n",
    "rm -rf TEMP && orthofinder -f analysis/orthodata/ -o TEMP && cp -r TEMP/Results_*/Orthogroups/ analysis\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c2a1cdd-a57c-4901-9d14-7bff652c41f4",
   "metadata": {},
   "source": [
    "# Reading orthogroups and cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "9e557a32-fcaa-492b-8744-7554e916cb99",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The orthogroup contains 7627 groups. Each sample has:\n",
      "Found 7533 groups in sample infantum in including 8152 genes\n",
      "Found 7551 groups in sample tarentolae in including 8261 genes\n",
      "\n",
      "Unassigned genes.\n",
      "Found 973 unassigned genes in sample Orthogroup\n",
      "Found 531 unassigned genes in sample infantum\n",
      "Found 442 unassigned genes in sample tarentolae\n"
     ]
    }
   ],
   "source": [
    "from sequana.orthofinder import OrthoGroups\n",
    "o = OrthoGroups(\"analysis/Orthogroups/\")\n",
    "o.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "94bfc06b-eb04-4ad2-bf72-06d3ad3b0b02",
   "metadata": {},
   "outputs": [],
   "source": [
    "ortho_groups = pd.read_csv(\"analysis/Orthogroups/Orthogroups.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "ff891ede-6f36-476f-b19f-c290e4fa70c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7457"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of orthogroups with both species\n",
    "len(ortho_groups.dropna())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "1975f84b-5ec5-4752-84e7-0f4c5be950e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of groups where infantum is missing : 94\n",
      "Number of groups where tarentolae is missing : 76\n"
     ]
    }
   ],
   "source": [
    "# number of groups where infantum is undefined (tarentolae only)\n",
    "print(f\"Number of groups where infantum is missing : {ortho_groups.isna()['infantum'].sum()}\")\n",
    "\n",
    "# number of groups where tarentolae is undefined (infantum only)\n",
    "print(f\"Number of groups where tarentolae is missing : {ortho_groups.isna()['tarentolae'].sum()}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "e78ad416-8cab-443a-843c-da175e0e43e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of tarentolae genes where infantum is missing : 342\n",
      "Number of infantum genes where tarentolae is missing : 353\n"
     ]
    }
   ],
   "source": [
    "S = sum([len(x.split(\",\")) for x in ortho_groups[ortho_groups.isna()['infantum']]['tarentolae']])\n",
    "print(f\"Number of tarentolae genes where infantum is missing : {S}\")\n",
    "S = sum([len(x.split(\",\")) for x in ortho_groups[ortho_groups.isna()['tarentolae']]['infantum']])\n",
    "print(f\"Number of infantum genes where tarentolae is missing : {S}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "07405e78-c10a-4cd4-8511-47b856004b5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let us drop the orthogroups with only one species.\n",
    "ortho_groups = ortho_groups.dropna()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e6396bb-43c4-46a3-a3d5-ee93206d398b",
   "metadata": {},
   "source": [
    "# Some groups are not homogenous.\n",
    "- case 1: for a given group, one species has several genes that are not on the same chromosome\n",
    "- case 2: the chromomosome for species 1 is different from species 2\n",
    "\n",
    "Those cases maybe biological but we want to group orthogroups by chromosome so, let us filter these cases"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93c1e6cc-fa73-4550-b49c-a7717a5454b1",
   "metadata": {},
   "source": [
    "## cleaning 1: non homogenous chromosomal gene location within a species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c774204b-62a8-4c53-b71a-9182c10ad638",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first, remove groups where chromosomes are not homogenous within one of the species\n",
    "indices_inf = []\n",
    "indices_tar = []\n",
    "for index in ortho_groups.index:    \n",
    "    # focus on infantum first\n",
    "    x = ortho_groups.loc[index,'infantum']\n",
    "    # chromosome is contain in the name\n",
    "    X = set([y.split(\"_\")[-1][0:2] for y in x.split(\",\")])\n",
    "    if len(X) != 1:\n",
    "        indices_inf.append(index)\n",
    "        #print(f\"infantum {index, X}\")\n",
    "\n",
    "\n",
    "    x = ortho_groups.loc[index,'tarentolae']\n",
    "    # chromosome is contain in the name\n",
    "    X = set([y.split(\"_\")[-1][0:2] for y in x.split(\",\")])\n",
    "    if len(X) != 1:\n",
    "        indices_tar.append(index)\n",
    "        #print(f\"tarentolae {index, X}\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "1f509b2e-71a5-4037-a189-cedef813520d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47 80 115\n"
     ]
    }
   ],
   "source": [
    "# 47 incoherent groups in infantum, 80 in tarentolae and 115 in total\n",
    "S1 = set(indices_inf)\n",
    "S2 = set(indices_tar)\n",
    "S3 = S1.union(S2)\n",
    "print(len(S1), len(S2), len(S3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "d49245f3-4646-4fd8-8d02-9a738b721f89",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove the groups where one of the species has a heterogenous chromosomal gene locations\n",
    "ortho_groups = ortho_groups.loc[[x for x in ortho_groups.index if x not in S3]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "eb2b4298-4495-4ab1-8127-f8f06037ceea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# let us add the chromosome of the species for each group\n",
    "# add chromosome\n",
    "chr_name_inf = []\n",
    "chr_name_tar = []\n",
    "for index in ortho_groups.index:    \n",
    "    x = ortho_groups.loc[index,'infantum']\n",
    "    X = set([y.split(\"_\")[-1][0:2] for y in x.split(\",\")])\n",
    "    chr_name_inf.append(list(X)[0])\n",
    "\n",
    "    x = ortho_groups.loc[index,'tarentolae']\n",
    "    X = set([y.split(\"_\")[-1][0:2] for y in x.split(\",\")])\n",
    "    chr_name_tar.append(list(X)[0])\n",
    "ortho_groups['chr_name_inf'] = chr_name_inf\n",
    "ortho_groups['chr_name_tar'] = chr_name_tar\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbe0aacb-deed-4b1e-9801-161de5fecbdc",
   "metadata": {},
   "source": [
    "# Cleaning 2: remove incoherent chromosome location between the 2 species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "e9278b32-2ece-483b-bfb5-9c92530205aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 25 incoherent chrom name\n"
     ]
    }
   ],
   "source": [
    "indices = []\n",
    "for index in ortho_groups.index:\n",
    "    x = ortho_groups.loc[index, 'chr_name_inf']\n",
    "    y = ortho_groups.loc[index, 'chr_name_tar']\n",
    "    if x!=y:\n",
    "        indices.append(index)\n",
    "print(f\"Found {len(indices)} incoherent chrom name\")\n",
    "ortho_groups = ortho_groups.loc[[x for x in ortho_groups.index if x not in indices]]\n",
    "# 25 incoherent chrom names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "1ea8645e-93e8-478c-8103-69fb0616230e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7317"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ortho_groups)\n",
    "# should be 7317 but regression in the notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a8d1e5a3-27fe-4a72-9b12-84faaf882651",
   "metadata": {},
   "outputs": [],
   "source": [
    "Nfinal = sum([len(x.split(',')) for x in ortho_groups['infantum'].values]) + sum([len(x.split(',')) for x in ortho_groups['tarentolae'].values])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "92bc1d6e-ceba-424f-b0b0-9fed4e8970ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original orthofinder assignement is 94.4035430806396%\n",
      "Original orthofinder assignement is 87.36339583572989%\n"
     ]
    }
   ],
   "source": [
    "N = len(GFF3(\"references/tarentolae.gff\").df.query(\"genetic_type=='gene'\")) + len(GFF3(\"references/infantum.gff\").df.query(\"genetic_type=='gene'\")) \n",
    "print(f\"Original orthofinder assignement is {16413 / N * 100}%\")\n",
    "final_assignation = Nfinal / N * 100\n",
    "print(f\"Original orthofinder assignement is {final_assignation}%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "675c2bdb-74d7-48f2-8617-c9f9c72a4438",
   "metadata": {},
   "outputs": [],
   "source": [
    "ortho_groups.to_csv(\"analysis//Orthogroups_cleaned.tsv\", sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68d1168e-b2fb-4892-aff0-2bd7f6d8321d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
